### DISCUSSION TITLE
ENV : clarification about building

### DISCUSSION THREAD

**Discussion by pubpub-zz** [py-pdf/pypdf#974] [Q&A]
@MartinThoma,
To prevent too many cosmetic commits can you please re-clarify what are the command line parameters and config file used during generation.
Also, do you have some reference/recommendation about codevod to propose the best coverage : I was expecting more pdf input files to cover coverage, but not sure it's working

**Comment by pubpub-zz** [comment]
converted to discussion

**Comment by MartinThoma** [comment]
Sure!

The simplest stylistic changes I make are done be [`pre-commit`](https://pre-commit.com/). You can do those automatically with every commit by running once

```
pip install pre-commit

# Within the PyPDF2 git root folder:
pre-commit install 

# it should show something like 'pre-commit installed at .git/hooks/pre-commit'
```

For existing PRs, you can run `pre-commit run --all`. Common changes are:

* Apply the black formatter
* Apply isort for sorting imports
* Ensure files end with a newline

**Comment by MartinThoma** [comment]
Simple, but not easily automatable:

* PEP8 naming: The flake8-plugin [`pep8-naming`](https://pypi.org/project/pep8-naming/) points out simple PEP8 issues, but we still have the functions/methods/parameters that got deprecated. So there are quite a lot of messages
* Private/public API: Keeping the public interface (anything that does not start with an underscore) small is desirable as we can easily change things. If we modify the behavior / name of anything of the public interface, we have to [go through the deprecation process](https://pypdf2.readthedocs.io/en/latest/dev/deprecations.html) to ensure we're not unexpectedly breaking other peoples code.

Complex:

* Code structure: I don't want functions / classes / modules to become too big. On the one hand, it makes them hard for me to understand. On the other hand, there might be more merge conflicts and the merge conflicts become more complicated.

**Comment by MartinThoma** [comment]
> do you have some reference/recommendation about codevod to propose the best coverage

On a more general note about testing: I'm currently downloading the Tika test dataset of 800,000 PDF files. Well, a part of it as my laptop doesn't have enough storage :sweat_smile: I plan to run simple operations over all of those files to ensure that we don't crash.

The other path I'm trying to go is to increase test coverage to 100% and kill all mutants found by the [mutmut](https://pypi.org/project/mutmut/) mutation tester ([intro to mutation testing](https://medium.com/analytics-vidhya/unit-testing-in-python-mutation-testing-7a70143180d8) if you're curious).

Increasing test coverage is typically also easier if you have smaller functions. Then you just test the function - you might not even need a PDF for that. Those test typically also run faster, but there is a higher risk that you're testing something that cannot happen / that the test itself is not so good. Mutation testing improves the quality of the tests.

**Comment by pubpub-zz** [comment]
@MartinThoma , @MasterOdin 
I don't know if I'm right, but black seems to be configured to  work with 80 char-long lines. I've always considered that this recommandation is obsolete with your high resolution/wide screens. Shouldn't we consider to reconfigure black to 120 ?

**Reply by MasterOdin** [reply]
[black is configured to use 88 char long lines](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#line-length), with flake8 configured to ignore line length errors (`E501`).

I personally like 88 (with trying to be around 80) as it allows me to open two files next to each other and see the entire line on my 16" display.

**Reply by MartinThoma** [reply]
I work on a 14-inch laptop (with super high resolution, but still) and I'm also reading code on my phone. I also like to view diff on Github.

The 88 char limit of black seems good to me; in some of my private projects I set the limit to 79 characters. It works mostly fine when you don't use Java-esque variable names and avoid too many nested layers (which you should due to mental complexity)

**Comment by xilopaint** [comment]
I would replace flake8 with Pylint. It takes some time to get rid of all the messages but it worth it in my opinion.

**Reply by MartinThoma** [reply]
There is no reason we have to decide between Flake8 or pylint. We can use both.

I haven't put pylint in the CI because it gives way too many messages at the moment.

**Comment by pubpub-zz** [comment]
May I suggest to restore label "Help Wanted" in github in order to try to get attention of other developpers on some issues?

**Reply by MartinThoma** [reply]
Can non-maintainers assign labels to (their) issues? Because I actually do want help for every single open issue :sweat_smile: 

To which issues should I then apply the "Help Wanted" label?

**Reply by MartinThoma** [reply]
I thought about adding something like `difficulty:easy` / `difficulty:medium` / `difficulty:hard` so that new contributors could pick better where to start. Also, I thought about leaving some easy ones intentionally open for new contributors.

**Reply by pubpub-zz** [reply]
#1091
your idea about difficulty sound good however do you think that the label will be reported at higher level like for "want help" ?
![image](https://user-images.githubusercontent.com/4083478/180598900-9e7671d9-db39-4a8b-b177-f63c9e222e52.png)

**Reply by MartinThoma** [reply]
The labels will probably not shown like this.

Are you suggesting to put the "help wanted" label on easy issues to encourage people to start contributing?

**Reply by pubpub-zz** [reply]
yes

**Reply by MartinThoma** [reply]
Makes sense to me - I'll activate that when I'm back home.

Thank you for the suggestion 😊

**Reply by MartinThoma** [reply]
I've added the "help wanted" back + added the topic "help-wanted" to the project: https://github.com/topics/help-wanted?l=python&o=desc&s=stars
